File: OpenFlix-tvOS/OpenFlix-tvOS/Models/Domain/Recording.swift
import Foundation

// MARK: - Recording Status

enum RecordingStatus: String, Codable {
    case scheduled
    case recording
    case completed
    case failed

    var displayName: String {
        switch self {
        case .scheduled: return "Scheduled"
        case .recording: return "Recording"
        case .completed: return "Completed"
        case .failed: return "Failed"
        }
    }

    var icon: String {
        switch self {
        case .scheduled: return "clock"
        case .recording: return "record.circle"
        case .completed: return "checkmark.circle.fill"
        case .failed: return "exclamationmark.circle.fill"
        }
    }
}

// MARK: - Recording

struct Recording: Identifiable, Hashable {
    let id: Int
    let title: String
    let subtitle: String?
    let description: String?
    let thumb: String?
    let art: String?
    let channelId: String?
    let channelName: String?
    let channelLogo: String?
    let startTime: Date
    let endTime: Date
    let duration: Int               // milliseconds
    let status: RecordingStatus
    let filePath: String?
    let fileSize: Int?
    let seasonNumber: Int?
    let episodeNumber: Int?
    let seriesRecord: Bool
    let seriesRuleId: Int?
    let genres: [String]
    let contentRating: String?
    let year: Int?
    let rating: Double?
    let isMovie: Bool
    let viewOffset: Int?            // milliseconds
    let commercials: [Commercial]
    let priority: Int

    // MARK: - Computed Properties

    var durationFormatted: String {
        String.formatDuration(milliseconds: duration)
    }

    var fileSizeFormatted: String? {
        guard let size = fileSize else { return nil }
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = [.useGB, .useMB]
        formatter.countStyle = .file
        return formatter.string(fromByteCount: Int64(size))
    }

    var dateFormatted: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: startTime)
    }

    var timeRangeFormatted: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"
        return "\(formatter.string(from: startTime)) - \(formatter.string(from: endTime))"
    }

    var episodeLabel: String? {
        guard let season = seasonNumber, let episode = episodeNumber else { return nil }
        return "S\(season) E\(episode)"
    }

    var fullTitle: String {
        if let label = episodeLabel, let subtitle = subtitle {
            return "\(title) - \(label) - \(subtitle)"
        } else if let subtitle = subtitle {
            return "\(title): \(subtitle)"
        }
        return title
    }

    var progressPercent: Double {
        guard duration > 0, let offset = viewOffset else { return 0 }
        return Double(offset) / Double(duration)
    }

    var isInProgress: Bool {
        progressPercent > 0 && progressPercent < 0.9
    }

    var isUpcoming: Bool {
        status == .scheduled && startTime > Date()
    }

    var isCurrentlyRecording: Bool {
        status == .recording
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Recording, rhs: Recording) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - Commercial

struct Commercial: Identifiable {
    var id: String { "\(start)-\(end)" }
    let start: Int          // milliseconds
    let end: Int            // milliseconds

    var duration: Int {
        end - start
    }

    var startSeconds: Int {
        start / 1000
    }

    var endSeconds: Int {
        end / 1000
    }
}

// MARK: - DTO Mapping

extension RecordingDTO {
    func toDomain() -> Recording {
        Recording(
            id: safeId,
            title: safeTitle,
            subtitle: subtitle,
            description: description ?? summary,
            thumb: thumb,
            art: art,
            channelId: channelId?.stringValue,
            channelName: channelName,
            channelLogo: channelLogo,
            startTime: startDate ?? Date(),
            endTime: endDate ?? Date(),
            duration: duration ?? 0,
            status: RecordingStatus(rawValue: safeStatus) ?? .scheduled,
            filePath: filePath,
            fileSize: fileSize,
            seasonNumber: seasonNumber,
            episodeNumber: episodeNumber,
            seriesRecord: seriesRecord ?? false,
            seriesRuleId: seriesRuleId,
            genres: genres?.split(separator: ",").map { String($0).trimmingCharacters(in: .whitespaces) } ?? [],
            contentRating: contentRating,
            year: year,
            rating: rating,
            isMovie: isMovie ?? false,
            viewOffset: viewOffset,
            commercials: commercials?.map { Commercial(start: $0.start, end: $0.end) } ?? [],
            priority: priority ?? 50
        )
    }
}

// MARK: - Series Rule

struct SeriesRule: Identifiable {
    let id: Int
    let title: String
    let channelId: String?
    let enabled: Bool
    let prePadding: Int
    let postPadding: Int
    let keepCount: Int
    let recordingCount: Int
}

extension SeriesRuleDTO {
    func toDomain() -> SeriesRule {
        SeriesRule(
            id: safeId,
            title: safeTitle,
            channelId: channelId?.stringValue,
            enabled: enabled ?? true,
            prePadding: prePadding ?? 0,
            postPadding: postPadding ?? 0,
            keepCount: keepCount ?? 0,
            recordingCount: recordingCount ?? 0
        )
    }
}

File: openflix-kmp/shared/src/commonMain/kotlin/com/openflix/domain/model/Recording.kt
package com.openflix.domain.model

enum class RecordingStatus(val value: String) {
    SCHEDULED("scheduled"),
    RECORDING("recording"),
    COMPLETED("completed"),
    FAILED("failed");

    val displayName: String get() = when (this) {
        SCHEDULED -> "Scheduled"
        RECORDING -> "Recording"
        COMPLETED -> "Completed"
        FAILED -> "Failed"
    }

    companion object {
        fun fromValue(value: String): RecordingStatus =
            entries.firstOrNull { it.value == value } ?: SCHEDULED
    }
}

data class Recording(
    val id: Int,
    val title: String,
    val subtitle: String? = null,
    val description: String? = null,
    val thumb: String? = null,
    val art: String? = null,
    val channelId: String? = null,
    val channelName: String? = null,
    val channelLogo: String? = null,
    val startTimeMs: Long,
    val endTimeMs: Long,
    val duration: Int,
    val status: RecordingStatus,
    val filePath: String? = null,
    val fileSize: Int? = null,
    val seasonNumber: Int? = null,
    val episodeNumber: Int? = null,
    val seriesRecord: Boolean = false,
    val seriesRuleId: Int? = null,
    val genres: List<String> = emptyList(),
    val contentRating: String? = null,
    val year: Int? = null,
    val rating: Double? = null,
    val isMovie: Boolean = false,
    val viewOffset: Int? = null,
    val commercials: List<Commercial> = emptyList(),
    val priority: Int = 50
) {
    val episodeLabel: String? get() {
        val s = seasonNumber ?: return null
        val e = episodeNumber ?: return null
        return "S$s E$e"
    }

    val fullTitle: String get() {
        val label = episodeLabel
        return when {
            label != null && subtitle != null -> "$title - $label - $subtitle"
            subtitle != null -> "$title: $subtitle"
            else -> title
        }
    }

    val progressPercent: Double get() {
        if (duration <= 0) return 0.0
        val offset = viewOffset ?: return 0.0
        return offset.toDouble() / duration.toDouble()
    }

    val isInProgress: Boolean get() = progressPercent > 0 && progressPercent < 0.9

    val isUpcoming: Boolean get() = status == RecordingStatus.SCHEDULED && startTimeMs > currentTimeMs()

    val isCurrentlyRecording: Boolean get() = status == RecordingStatus.RECORDING
}

data class Commercial(
    val start: Int,
    val end: Int
) {
    val duration: Int get() = end - start
    val startSeconds: Int get() = start / 1000
    val endSeconds: Int get() = end / 1000
}

data class SeriesRule(
    val id: Int,
    val title: String,
    val channelId: String? = null,
    val enabled: Boolean = true,
    val prePadding: Int = 0,
    val postPadding: Int = 0,
    val keepCount: Int = 0,
    val recordingCount: Int = 0
)
